src/core/thread/mle.cpp

```diff
@@ -1223,12 +1226,39 @@ otError Mle::AppendAddressRegistration(Message &aMessage)
 
         SuccessOrExit(error = aMessage.Append(&entry, entry.GetLength()));
         length += entry.GetLength();
+        counter++;
+        // only continue to append if there is available entry.
+        VerifyOrExit(counter < Child::kMaxIp6AddressPerChild);
```

Comment:
There are some points to send all addresses. However there are something we may need think more.
For example,  after parent reboot, we need to ensure the Child Update Response from Child to fit in one frame as Thread specification requires all fragmented MLE message to have MAC security enabled. Unlimited entries for AddressRegistrationTLV in MLE message from Child would bring risks for packet fragments, may causing re-sync failing.
A rough estimation is that at most, only (4 compressed) or (2 compressed + 1 uncompressed) addresses in AddressRegistrationTLV could help ensure above Child Update Response to fit in one frame.

@jwhui , maybe the most safe way for the parent to re-sync with its MTD children after reboot is no `AddressRegistrationTLV request` in Child Update Request, letting the re-sync Child Update Request/Response exchange only serves for link re-establishment, and the child could explicitly send Child Update Request to register its addresses afterwards. Thoughts?

Reply:
Ahh, okay i didn't think about this. Thank you for detailed analysis and proposal.

First of all, regardless of the solution we will accept, we shouldn't use `Child::kMaxIp6AddressPerChild` constant, when we operating as a SED, i think its very misleading, due to the reasons given above. I guess we can create a new constant that better describes its usage.

Although i didn't think that through, i like very much your proposal, to postpone address registration after link is re-established. I believe Thread specification requires us to always include at least ML-EID address (in Child ID Request).

> MTDs MUST include an Address Registration TLV containing their ML-EID along with any other addresses they wish to register with their parent.

However, in case of re-synchronization, we need to take into consideration the fact, that parent may remove (?) our addresses that we previously registered with him, which may lead to packet drop from indirect queue.

Reply:
Agree to use a new constant way. Thanks.
Additionally, filed [SPEC-751](https://threadgroup.atlassian.net/browse/SPEC-751) to further discuss the secured MLE fragments issues during attach/re-sync process due to many entries in AddressRegistration TLV. Please help to share your more concerns there. Thanks.